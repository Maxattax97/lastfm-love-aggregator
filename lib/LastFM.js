const _ = require('lodash');
const got = require('got');
const cheerio = require('cheerio');
const {default: PQueue} = require('p-queue');

class LastFM {
    constructor(opt) {
        opt = _.defaultsDeep(opt, {
            root: 'http://ws.audioscrobbler.com/2.0/'
        });

        this.root = opt.root;
        this.apiKey = opt.apiKey;
        this.sharedSecret = opt.sharedSecret;

        this.apiQueue = new PQueue({
            intervalCap: 1,
            interval: 1000
        });

        this.scrapeQueue = new PQueue({
            intervalCap: 2,
            interval: 800
        });
    }

    async apiRequest(opt) {
        opt = _.defaultsDeep(opt, {
            params: {
                api_key: this.apiKey,
                format: 'json',
            }
        });

        return this.apiQueue.add(() => got(this.root, { searchParams: opt.params }).json());
    }

    // TODO: Paginate this, got has a feature for this.
    async userGetLovedTracks(opt)
    {
        opt = _.defaultsDeep(opt, {
            user: 'Maxattax97',
            limit: 50,
            page: 1
        });

        return this.apiRequest({ params: {
            method: 'user.getLovedTracks',
            user: opt.user,
            limit: opt.limit,
            page: opt.page
        }});
    }

    async scrapeSong(opt)
    {
        opt = _.defaultsDeep(opt, {
            title: 'Untitled',
            artist: 'Unknown',
            url: 'https://www.last.fm/music/Kali+Uchis/_/Coming+Home+(Interlude)'
        });

        const page = await this.scrapeQueue.add(() => got(opt.url));
        const $ = cheerio.load(page.body);
        const youtubeUrl = $('#track-page-video-playlink').attr("href"); // undefined if it doesn't exist.

        return {
            title: opt.title,
            artist: opt.artist,
            lastfmUrl: opt.url,
            youtubeUrl: youtubeUrl
        }
    }

    parseTracks(tracks)
    {
        return _.map(tracks, (track) => {
            return {
                title: track.name,
                artist: track.artist.name,
                url: track.url
            }
        });
    }
}

module.exports = LastFM;
