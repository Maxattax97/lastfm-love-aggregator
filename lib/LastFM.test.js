/* eslint-env jest */

import fs from 'fs-extra';
// import { jest } from '@jest/globals';
import _ from 'lodash';
import LastFM from './LastFM.js';

describe('LastFM', () => {
  let lfm = null;
  let hasApiKey = false;

  beforeAll(async () => {
    let configuration = {};
    try {
      configuration = JSON.parse(
        (await fs.readFile('./api-keys.json')).toString(),
      );
      hasApiKey = true;
    } catch (err) {
      console.log('no API key provided, skipping those tests');
      configuration.apiKey = 'REDACTED';
      configuration.sharedSecret = 'REDACTED';
    }

    lfm = new LastFM({
      apiKey: configuration.apiKey,
      sharedSecret: configuration.sharedSecret,
    });
  });

  it('should get and parse loved LastFM tracks', async () => {
    if (!hasApiKey) {
      console.log('Skipping this test because no API key was provided');
      return;
    }

    const lovedResponse = await lfm.userGetLovedTracks({
      user: 'Maxattax97',
      limit: 5,
    });

    const loved = LastFM.parseTracks(lovedResponse.lovedtracks.track);
    expect(loved.length).toEqual(5);

    const firstTrack = loved[0];
    expect(typeof firstTrack.title).toBe('string');
    expect(typeof firstTrack.mbid).toBe('string');
    expect(typeof firstTrack.artist).toBe('string');
    expect(typeof firstTrack.artistMbid).toBe('string');
    expect(_.startsWith(firstTrack.url, 'https://www.last.fm/music/')).toBe(
      true,
    );
  });

  it('should scrape [YOUR NAME by DVRST]', async () => {
    const demoTrack = {
      title: 'YOUR NAME',
      artist: 'DVRST',
      artistMbid: 'b04b7c24-59d2-47f8-8ca8-bcf6f6d106ed',
      url: 'https://www.last.fm/music/DVRST/_/YOUR+NAME',
      mbid: '750ee163-e687-445e-a62b-10cf2f75dd83',
    };

    const ytScrape = await lfm.scrapeSong(demoTrack);

    expect(ytScrape).toMatchSnapshot();
  });
});
