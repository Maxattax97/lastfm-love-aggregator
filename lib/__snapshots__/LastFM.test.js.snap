// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastFM should scrape [YOUR NAME by DVRST] 1`] = `
{
  "artist": "DVRST",
  "artistMbid": "b04b7c24-59d2-47f8-8ca8-bcf6f6d106ed",
  "lastfmUrl": "https://www.last.fm/music/DVRST/_/YOUR+NAME",
  "mbid": "750ee163-e687-445e-a62b-10cf2f75dd83",
  "title": "YOUR NAME",
  "youtubeUrl": "https://www.youtube.com/watch?v=W9GXqCBAZb0",
}
`;
